File: python/option.py
from monad import Monad

class Option(Monad):
  def pure(self, x):
    return Some(x)

  def flat_map(self, f):
    if self.defined:
      return f(self.value)
    else:
      return nil


class Some(Option):
  def __init__(self, value):
    self.value = value
    self.defined = True

class Nil(Option):
  def __init__(self):
    self.value = None
    self.defined = False

nil = Nil()

# print(Some(1).map(lambda x: x + 1).value)
# print(Some(1).flat_map(lambda x: nil).value)
# print(Some(1).flat_map(lambda x: Some(x + 3)).value)
# print(nil.map(lambda x: x + 1).defined)

File: ruby/option.rb
require './monad'

class Option < Monad

  attr_accessor :defined, :value

  # pure :: a -> Option a
  def self.pure(x)
    Some.new(x)
  end

  # pure :: a -> Option a
  def pure(x)
    Some.new(x)
  end
  
  # flatMap :: # Option a -> (a -> Option b) -> Option b
  def flat_map(f)
    if defined
      f.call(value)
    else
      $none
    end
  end
end

class Some < Option
  def initialize(value)
    @value = value
    @defined = true
  end
end

class None < Option
  def initialize()
    @defined = false
  end
end

$none = None.new()

# puts Some.new(1).map(-> (x) { x + 1 }).value
# puts Some.new(1).flat_map(-> (x) { Some.new(x + 1) }).value
# puts Some.new(1).flat_map(-> (x) { $none }).defined
# puts $none.flat_map(-> (x) { Some.new(3) }).defined
